src/compas_fab/robots/robot.py

```diff
@@ -47,15 +48,43 @@ class Robot(object):
         Dictionary mapping planning groups to the tool currently attached to them, if any.
     """
 
-    def __init__(self, model, artist=None, semantics=None, client=None):
+    def __init__(self, model=None, artist=None, semantics=None, client=None):
+        super(Robot, self).__init__()
+        # These attributes have to be initiated first,
+        # because they are used in the setters of the other attributes
         self._scale_factor = 1.0
-        self.model = model
         self._attached_tools = {}  # { planning_group_name: robots.tool.Tool }
+        self._current_ik = {"request_id": None, "solutions": None}
+
+        self.model = model
         self.artist = artist
         self.semantics = semantics
         self.client = client
         self.attributes = {}
-        self._current_ik = {"request_id": None, "solutions": None}
+
+    @property
+    def data(self):
+        data = {
+            "_scale_factor": self._scale_factor,
+            "_attached_tools": self._attached_tools,
+            "_current_ik": self._current_ik,
+            "model": self.model.data,
+            "semantics": self.semantics,
+            "attributes": self.attributes,
+            # The following attributes cannnot be serizlied
+            # "artist": self.artist.data if self.artist else None,
+            # "client": self.client.data if self.client else None,
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self._scale_factor = data.get("_scale_factor", 1.0)
+        self._attached_tools = data.get("_attached_tools", {})
+        self._current_ik = data.get("_current_ik", {"request_id": None, "solutions": None})
```

Comment:
mmh is this to stay compatible with previously serialized `Robot`s? But it didn't implement the `Data` interface before, so there shouldn't be any, right?

You could also simply omit these from the `Data` dictionary if they take default values, `__init__` is ran first anyways. Otherwise you can assume they're there, because they're put there by the getter.